File: pset1/credit.c
/**
 * PSet 1.3b - Credit
 * Written by Sudipto Ghosh for CS50
 */

#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    // declare essential variables
    int cardType = 0;
    long cardNumber = 0l;
    
    // prompt user for card number
    // and assign it to 'cardNumber'
    do
    {
        cardNumber = get_long("Number: ");
    }
    while (cardNumber <= 0);  
    
    // calculate number of digits
    int digits = 0;
    long validation = cardNumber;
    while (validation > 0)
    {
        validation /= 10;
        digits++;
    }
    
    // find out checksum digits * 2
    int checksumDigits[8] = {0};
    for (int i = 0, j = 1; i < 8; i++, j += 2)
    {
        checksumDigits[i] = (int)(cardNumber / (long)(pow(10, j)) % 10 * 2);        
    }
    
    // calculate sum of digits of the products
    int sum = 0;
    for (int i = 0;  i < 8; i++)
    {
        for (int j = 0; j < 2; j++, checksumDigits[i] /= 10)
        {
            sum += checksumDigits[i] % 10;
        }
    }
    
    // add remaining digits of card number to sum
    for (int i = 0, j = 0; i < 8; i++, j += 2)
    {
        sum += (int)(cardNumber / (long)(pow(10, j)) % 10);        
    }
    
    // check one's digit of the total
    if (sum % 10 != 0)
    {
        cardType = -1;
    }
    else
    {
        // check starting digits of credit cards
        switch (digits)
        {
            case 13:
                if (cardNumber / (long)(pow(10, 12)) == 4)
                {
                    cardType = 1;
                }
                else
                {
                    cardType = -1;
                }
                break;
            case 15:
                if (cardNumber / (long)(pow(10, 13)) == 34 ||
                    cardNumber / (long)(pow(10, 13)) == 37)
                {
                    cardType = 3;
                }
                else
                {
                    cardType = -1;
                }
                break;
            case 16:
                if (cardNumber / (long)(pow(10, 15)) == 4)
                {
                    cardType = 1;
                }
                else if (cardNumber / (long)(pow(10, 14)) == 51 ||
                         cardNumber / (long)(pow(10, 14)) == 52 ||
                         cardNumber / (long)(pow(10, 14)) == 53 ||
                         cardNumber / (long)(pow(10, 14)) == 54 ||
                         cardNumber / (long)(pow(10, 14)) == 55)
                {
                    cardType = 2;
                }
                else
                {
                    cardType = -1;
                }
                break;
            default:
                cardType = -1;
                break;
        }
    }
    
    // print card type
    switch (cardType)
    {
        case 1:
            printf("VISA\n");
            break; 
        case 2:
            printf("MASTERCARD\n");
            break;
        case 3:
            printf("AMEX\n");
            break;
        case -1:
            printf("INVALID\n");
            break;
    }
    
    return 0;
}

File: pset6/credit/credit.py
"""
  PSet 6 - Sentimental/Credit
  Written by Sudipto Ghosh for CS50
"""

from cs50 import get_int

# get card number from user input
while True:
    cardNumber = get_int("Number: ")
    if cardNumber > 0:
        break

# calculate number of digits
digits = len(str(cardNumber))

# find out checksum digits * 2
checksumDigits = []
for i in range(1, 17, 2):
    checksumDigits.append((cardNumber // 10 ** i) % 10 * 2)

# calculate sum of checksum digits
sum = 0
for i in checksumDigits:
    for j in range(2):
        sum += i % 10
        i //= 10

# add remaining digits to sum
for j in range(0, 16, 2):
    sum += (cardNumber // 10 ** j) % 10

# determine and print card type
if sum % 10 == 0:
    first = int(str(cardNumber)[0])
    firstTwo = int(str(cardNumber)[0]) * 10 + int(str(cardNumber)[1])

    if digits == 13:
        if first == 4:
            print("VISA")
        else:
            print("INVALID")
    elif digits == 15:

        if (firstTwo == 34 or firstTwo == 37):
            print("AMEX")
        else:
            print("INVALID")
    elif digits == 16:
        if first == 4:
            print("VISA")
        elif (firstTwo == 51 or
              firstTwo == 52 or
              firstTwo == 53 or
              firstTwo == 54 or
              firstTwo == 55):
            print("MASTERCARD")
        else:
            print("INVALID")
else:
    print("INVALID")
